src/Microsoft.AspNetCore.Mvc.TagHelpers/CacheTagHelperBase.cs

```diff
@@ -0,0 +1,307 @@
+        /// </summary>
+        [HtmlAttributeNotBound]
+        [ViewContext]
+        public ViewContext ViewContext { get; set; }
+
+        /// <summary>
+        /// Gets or sets a <see cref="string" /> to vary the cached result by.
+        /// </summary>
+        [HtmlAttributeName(VaryByAttributeName)]
+        public string VaryBy { get; set; }
+
+        /// <summary>
+        /// Gets or sets the name of a HTTP request header to vary the cached result by.
+        /// </summary>
+        [HtmlAttributeName(VaryByHeaderAttributeName)]
+        public string VaryByHeader { get; set; }
+
+        /// <summary>
+        /// Gets or sets a comma-delimited set of query parameters to vary the cached result by.
+        /// </summary>
+        [HtmlAttributeName(VaryByQueryAttributeName)]
+        public string VaryByQuery { get; set; }
+
+        /// <summary>
+        /// Gets or sets a comma-delimited set of route data parameters to vary the cached result by.
+        /// </summary>
+        [HtmlAttributeName(VaryByRouteAttributeName)]
+        public string VaryByRoute { get; set; }
+
+        /// <summary>
+        /// Gets or sets a comma-delimited set of cookie names to vary the cached result by.
+        /// </summary>
+        [HtmlAttributeName(VaryByCookieAttributeName)]
+        public string VaryByCookie { get; set; }
+
+        /// <summary>
+        /// Gets or sets a value that determines if the cached result is to be varied by the Identity for the logged in
+        /// <see cref="Http.HttpContext.User"/>.
+        /// </summary>
+        [HtmlAttributeName(VaryByUserAttributeName)]
+        public bool VaryByUser { get; set; }
+
+        /// <summary>
+        /// Gets or sets the exact <see cref="DateTimeOffset"/> the cache entry should be evicted.
+        /// </summary>
+        [HtmlAttributeName(ExpiresOnAttributeName)]
+        public DateTimeOffset? ExpiresOn { get; set; }
+
+        /// <summary>
+        /// Gets or sets the duration, from the time the cache entry was added, when it should be evicted.
+        /// </summary>
+        [HtmlAttributeName(ExpiresAfterAttributeName)]
+        public TimeSpan? ExpiresAfter { get; set; }
+
+        /// <summary>
+        /// Gets or sets the duration from last access that the cache entry should be evicted.
+        /// </summary>
+        [HtmlAttributeName(ExpiresSlidingAttributeName)]
+        public TimeSpan? ExpiresSliding { get; set; }
+
+        /// <summary>
+        /// Gets or sets the value which determines if the tag helper is enabled or not.
+        /// </summary>
+        [HtmlAttributeName(EnabledAttributeName)]
+        public bool Enabled { get; set; } = true;
+
+        protected abstract string GetUniqueId(TagHelperContext context);
+
+        protected abstract string GetKeyPrefix(TagHelperContext context);
+
+        // Internal for unit testing
+        internal string GenerateKey(TagHelperContext context)
```

Comment:
Could we make the storage support non-string keys? It'd be nice if we can change this code to use a type with `IEquatable<>` to avoid the string allocations.